apps/site/lib/site/realtime_schedule.ex

```diff
@@ -0,0 +1,222 @@
+  ]
+
+  @type route_with_patterns_t :: {
+          Stop.id_t(),
+          Route.t(),
+          [RoutePattern.t()]
+        }
+
+  @type route_pattern_name_t :: String.t()
+
+  @spec stop_data([Stop.id_t()], DateTime.t(), Keyword.t()) :: [map]
+  def stop_data(stop_ids, now, opts \\ []) do
+    opts = Keyword.merge(@default_opts, opts)
+    stops_fn = Keyword.fetch!(opts, :stops_fn)
+    routes_fn = Keyword.fetch!(opts, :routes_fn)
+    predictions_fn = Keyword.fetch!(opts, :predictions_fn)
+    schedules_fn = Keyword.fetch!(opts, :schedules_fn)
+
+    # stage 1, get stops and routes
+    stops_task = Task.async(fn -> get_stops(stop_ids, stops_fn) end)
+    routes_task = Task.async(fn -> get_routes(stop_ids, routes_fn) end)
+    [stops, route_with_patterns] = Enum.map([stops_task, routes_task], &Task.await/1)
+
+    # stage 2, get predictions and schedules
+    predictions_task =
+      Task.async(fn -> get_predictions(route_with_patterns, now, predictions_fn) end)
+
+    schedules_task = Task.async(fn -> get_schedules(route_with_patterns, now, schedules_fn) end)
+
+    [predictions, schedules] =
+      Enum.map([predictions_task, schedules_task], &Task.await(&1, @long_timeout))
+
+    build_output(stops, route_with_patterns, schedules, predictions)
+  end
+
+  @spec get_stops([Stop.id_t()], fun()) :: map
+  defp get_stops(stop_ids, stops_fn) do
+    stop_ids
+    |> Enum.map(
+      &Task.async(fn ->
+        {&1, &1 |> stops_fn.() |> Map.take([:id, :name, :accessibility, :address])}
+      end)
+    )
+    |> Enum.into(%{}, &Task.await/1)
+  end
+
+  @spec get_routes([Stop.id_t()], fun()) :: [route_with_patterns_t]
+  defp get_routes(stop_ids, routes_fn) do
+    stop_ids
+    |> Enum.map(
+      &Task.async(fn ->
+        &1
+        |> routes_fn.()
+        |> Enum.map(fn {route, route_patterns} ->
+          {&1, Route.to_json_safe(route), route_patterns}
+        end)
+      end)
+    )
+    |> Enum.map(&Task.await/1)
+    |> List.flatten()
+  end
+
+  @spec get_predictions([route_with_patterns_t], DateTime.t(), fun()) :: map
+  defp get_predictions(route_with_patterns, now, predictions_fn) do
+    route_with_patterns
+    |> Enum.map(fn {stop_id, _route, route_patterns} ->
+      Task.async(fn ->
+        do_get_predictions(stop_id, route_patterns, now, predictions_fn)
+      end)
+    end)
+    |> Enum.flat_map(&Task.await/1)
+    |> Enum.into(%{})
+  end
+
+  @spec do_get_predictions(Stop.id_t(), [RoutePattern.t()], DateTime.t(), fun()) :: [
+          {
+            route_pattern_name_t,
+            [Prediction.t()]
+          }
+        ]
+  defp do_get_predictions(stop_id, route_patterns, now, predictions_fn) do
+    route_patterns
+    |> Enum.map(fn route_pattern ->
+      Task.async(fn ->
+        next_two_predictions =
+          [
+            stop: stop_id,
+            route_pattern: route_pattern.id,
+            min_time: now,
+            sort: "time",
+            "page[limit]": 2
```

Comment:
this `2` would be better in a module attribute in case we ever want to change it